Add tutorial "Using Stamped datatypes with tf2_ros::MessageFilter" (#1929)
ros2/ros2_documentation: source/Tutorials/Tf2/Using-Stamped-Datatypes-With-Tf2-Ros-MessageFilter.rst

.. _UsingStampedDatatypesWithTf2RosMessageFilter:

Using stamped datatypes with tf2_ros::MessageFilter
===================================================

**Goal:** Learn how to use tf2_ros::MessageFilter to process stamped datatypes.

**Tutorial level:** Intermediate

**Time:** 10 minutes

.. contents:: Contents
   :depth: 3
   :local:

Background
----------

This tutorial explains how to use sensor data with tf2. Some real-world examples of sensor data are:

    * cameras, both mono and stereo

    * laser scans

Suppose that a new turtle named ``turtle3`` is created and it doesn't have good odometry, but there is an overhead camera tracking its position and publishing it as a ``PointStamped`` message in relation to the ``world`` frame.

``turtle1`` wants to know where ``turtle3`` is compared to itself.

> To do this ``turtle1`` must listen to the topic where ``turtle3``'s pose is being published, wait until transforms into the desired frame are ready, and then do it's operations. To make this easier the ``tf2_ros::MessageFilter`` is very useful. The ``tf2_ros::MessageFilter`` will take a subscription to any ROS 2 message with a header and cache it until it is possible to transform it into the target frame.

Reviewer: Here and elsewhere in this document, please do one sentence per-line.  That just makes it easier to make edits in the future.

Author: Yes, I did it just now, and now in the file each line only has one sentence.

Setting the example
-------------------

1 Write the broadcaster node of PointStamped messages
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For this tutorial we will set up a demo application which has a node (in Python) to broadcast the ``PointStamped`` messages of ``turtle3``.

First, let's create the source file.

Go to the ``learning_tf2_py`` :ref:`package <WritingATf2StaticBroadcasterPy>` we created in the previous tutorial.
Inside the ``src/learning_tf2_py/learning_tf2_py`` directory download the example sensor message broadcaster code by entering the following command:

.. tabs::

    .. group-tab:: Linux

        .. code-block:: console

            wget https://raw.githubusercontent.com/ros/geometry_tutorials/ros2/turtle_tf2_py/turtle_tf2_py/turtle_tf2_message_broadcaster.py

    .. group-tab:: macOS

        .. code-block:: console

            wget https://raw.githubusercontent.com/ros/geometry_tutorials/ros2/turtle_tf2_py/turtle_tf2_py/turtle_tf2_message_broadcaster.py

    .. group-tab:: Windows

        In a Windows command line prompt:

        .. code-block:: console

>                 curl -sk https://raw.githubusercontent.com/ros/geometry_tutorials/ros2/turtle_tf2_py/turtle_tf2_py/turtle_tf2_message_broadcaster.py -o turtle_tf2_message_broadcaster.py

Reviewer: Please make the indentation consistent with above, that is, get rid of 4 spaces here.  Same for the second curl line down below.

Author: Ok, I checked and modified all the indentation consistent in my tutorial file.

               ]
           ),
           Node(
               package='turtle_tf2_py',
               executable='turtle_tf2_message_broadcaster',
               name='message_broadcaster',
           ),
       ])

1.3 Add an entry point and build the pakcage
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Don't forget to add the executable in the ``setup.py`` file of the package:

.. code-block:: python

   'console_scripts': [
       ...
       'turtle_tf2_message_broadcaster = learning_tf2_py.turtle_tf2_message_broadcaster:main',
   ],

And then we can build the package:

.. tabs::

  .. group-tab:: Linux

    .. code-block:: console

       colcon build --packages-select learning_tf2_py

  .. group-tab:: macOS

    .. code-block:: console

       colcon build --packages-select learning_tf2_py

  .. group-tab:: Windows

    .. code-block:: console

       colcon build --merge-install --packages-select learning_tf2_py

2 Writing the message filter/listener node
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Now, to get the streaming ``PointStamped`` data of ``turtle3`` in the frame of ``turtle1`` reliably, we will create the source file at first.

> Go to the ``learning_tf2_cpp`` :ref:`package <WritingATf2StaticBroadcasterCpp>` we created in the previous tutorial.Inside the ``src/learning_tf2_cpp/src`` directory download file ``turtle_tf2_message_filter.cpp`` by entering the following command:

Reviewer 2: ```suggestion
Go to the ``learning_tf2_cpp`` :ref:`package <WritingATf2StaticBroadcasterCpp>` we created in the previous tutorial.
Inside the ``src/learning_tf2_cpp/src`` directory download file ``turtle_tf2_message_filter.cpp`` by entering the following command:
```

Author: Ok, I changed that line into two paragraphs.

.. _UsingStampedDatatypesWithTf2RosMessageFilter:

Using stamped datatypes with tf2_ros::MessageFilter
===================================================

**Goal:** Learn how to use ``tf2_ros::MessageFilter`` to process stamped datatypes.

**Tutorial level:** Advanced

**Time:** 10 minutes

.. contents:: Contents
   :depth: 3
   :local:

Background
----------

This tutorial explains how to use sensor data with tf2. Some real-world examples of sensor data are:

    * cameras, both mono and stereo

    * laser scans

Suppose that a new turtle named ``turtle3`` is created and it doesn't have good odometry, but there is an overhead camera tracking its position and publishing it as a ``PointStamped`` message in relation to the ``world`` frame.

``turtle1`` wants to know where ``turtle3`` is compared to itself.

> To do this ``turtle1`` must listen to the topic where ``turtle3``'s pose is being published, wait until transforms into the desired frame are ready, and then do its operations. To make this easier the ``tf2_ros::MessageFilter`` is very useful. The ``tf2_ros::MessageFilter`` will take a subscription to any ROS 2 message with a header and cache it until it is possible to transform it into the target frame.

Reviewer: ```suggestion
To do this ``turtle1`` must listen to the topic where ``turtle3``'s pose is being published, wait until transforms into the desired frame are ready, and then do its operations.
To make this easier the ``tf2_ros::MessageFilter`` is very useful.
The ``tf2_ros::MessageFilter`` will take a subscription to any ROS 2 message with a header and cache it until it is possible to transform it into the target frame.
```

Author: Ok, I split this paragraph into 3 paragraphs.

> Setting the example

Reviewer: ```suggestion
Setting up the example
```

Author: Yes, you're right, I should use "setting up" other than "setting". I changed that.

               else:
                   # Check if the service is ready
                   self.get_logger().info('Service is not ready')

           if self.turtle_pose_cansubscribe:
               self.vel_pub = self.create_publisher(Twist, 'turtle3/cmd_vel', 10)
               self.sub = self.create_subscription(Pose, 'turtle3/pose', self.handle_turtle_pose, 10)
               self.pub = self.create_publisher(PointStamped, 'turtle3/turtle_point_stamped', 10)

       def handle_turtle_pose(self, msg):
           vel_msg = Twist()
           vel_msg.linear.x = 1.0
           vel_msg.angular.z = 1.0
           self.vel_pub.publish(vel_msg)

           ps = PointStamped()
           ps.header.stamp = self.get_clock().now().to_msg()
           ps.header.frame_id = 'world'
           ps.point.x = msg.x
           ps.point.y = msg.y
           ps.point.z = 0.0
           self.pub.publish(ps)

   def main():
       rclpy.init()
       node = PointPublisher()
       try:
           rclpy.spin(node)
       except KeyboardInterrupt:
           pass

       rclpy.shutdown()

1.1 Examine the code
~~~~~~~~~~~~~~~~~~~~

Now let's take a look at the code.

> Firstly, in the ``on_timer`` callback function, we spawn the ``turtle3`` by asynchronously calling the ``Spawn`` service of ``turtlesim``, and initialize its position at (4, 2, 0), when the turtle spawning service is ready.

Reviewer: ```suggestion
First, in the ``on_timer`` callback function, we spawn the ``turtle3`` by asynchronously calling the ``Spawn`` service of ``turtlesim``, and initialize its position at (4, 2, 0), when the turtle spawning service is ready.
```

Author: Ok, the same error, I modified it.

.. code-block:: python

   # Initialize request with turtle name and coordinates
   # Note that x, y and theta are defined as floats in turtlesim/srv/Spawn
   request = Spawn.Request()
   request.name = 'turtle3'
   request.x = float(4)
   request.y = float(2)
   request.theta = float(0)
   Call request
   self.result = self.spawner.call_async(request)

Afterward, the node publishes the topic ``turtle3/cmd_vel``, topic ``turtle3/turtle_point_stamped``, and subscribes to topic ``turtle3/pose`` and runs callback function ``handle_turtle_pose`` on every incoming message.

.. code-block:: python

   self.vel_pub = self.create_publisher(Twist, '/turtle3/cmd_vel', 10)
   self.sub = self.create_subscription(Pose, '/turtle3/pose', self.handle_turtle_pose, 10)
   self.pub = self.create_publisher(PointStamped, '/turtle3/turtle_point_stamped', 10)

> Finally, in the callback function ``handle_turtle_pose``, we initialize the ``Twist`` messages of ``turtle3`` and publish them, which will make the ``turtle3`` move along a circle. Then we fill up the ``PointStamped`` messages of ``turtle3`` with incoming ``Pose`` messages and publish them.

Reviewer: ```suggestion
Finally, in the callback function ``handle_turtle_pose``, we initialize the ``Twist`` messages of ``turtle3`` and publish them, which will make the ``turtle3`` move along a circle.
Then we fill up the ``PointStamped`` messages of ``turtle3`` with incoming ``Pose`` messages and publish them.
```

Author: I split it into 2 paragraphs.

               executable='turtle_tf2_broadcaster',
               name='broadcaster2',
               parameters=[
                   {'turtlename': 'turtle3'}
               ]
           ),
           Node(
               package='turtle_tf2_py',
               executable='turtle_tf2_message_broadcaster',
               name='message_broadcaster',
           ),
       ])

1.3 Add an entry point and build the pakcage
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Don't forget to add the executable in the ``setup.py`` file of the package:

.. code-block:: python

   'console_scripts': [
       ...
       'turtle_tf2_message_broadcaster = learning_tf2_py.turtle_tf2_message_broadcaster:main',
   ],

And then we can build the package:

.. tabs::

  .. group-tab:: Linux

    .. code-block:: console

       colcon build --packages-select learning_tf2_py

  .. group-tab:: macOS

    .. code-block:: console

       colcon build --packages-select learning_tf2_py

  .. group-tab:: Windows

    .. code-block:: console

       colcon build --merge-install --packages-select learning_tf2_py

2 Writing the message filter/listener node
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Now, to get the streaming ``PointStamped`` data of ``turtle3`` in the frame of ``turtle1`` reliably, we will create the source file at first.

Reviewer: I'm not quite sure what this sentence is saying.  Is it that we are now creating the source file for `PointStamped` data, or something else?  Clarification here would be useful.

Author: Yes, I clarified as following:
![image](https://user-images.githubusercontent.com/89616578/135030243-d21f8ffb-b150-4e88-b52e-30dd9e861be3.png)

     }

   private:
     void msgCallback(const geometry_msgs::msg::PointStamped::SharedPtr point_ptr)
     {
       geometry_msgs::msg::PointStamped point_out;
       try {
         tf2_buffer_->transform(*point_ptr, point_out, target_frame_);
         RCLCPP_INFO(
           this->get_logger(), "Point of turtle3 in frame of turtle1: x:%f y:%f z:%f\n",
           point_out.point.x,
           point_out.point.y,
           point_out.point.z);
       } catch (tf2::TransformException & ex) {
         RCLCPP_WARN(
           // Print exception which was caught
           this->get_logger(), "Failure %s\n", ex.what());
       }
     }
     std::string target_frame_;
     std::shared_ptr<tf2_ros::Buffer> tf2_buffer_;
     std::shared_ptr<tf2_ros::TransformListener> tf2_listener_;
     message_filters::Subscriber<geometry_msgs::msg::PointStamped> point_sub_;
     std::shared_ptr<tf2_ros::MessageFilter<geometry_msgs::msg::PointStamped>> tf2_filter_;
   };

   int main(int argc, char * argv[])
   {
     rclcpp::init(argc, argv);
     rclcpp::spin(std::make_shared<PoseDrawer>());
     rclcpp::shutdown();
     return 0;
   }

2.1 Examine the code
~~~~~~~~~~~~~~~~~~~~

> Firstly, you must include the ``tf2_ros::MessageFilter`` headers from the ``tf2_ros`` package. As well as the previously used ``tf2`` and ``ros2`` related headers.

Reviewer: ```suggestion
First, you must include the ``tf2_ros::MessageFilter`` headers from the ``tf2_ros`` package, as well as the previously used ``tf2`` and ``ros2`` related headers.
```

Author: I changed these errors.

.. code-block:: C++

   #include <geometry_msgs/msg/point_stamped.hpp>
   #include <message_filters/subscriber.h>

   #include <rclcpp/rclcpp.hpp>
   #include <tf2_ros/buffer.h>
   #include <tf2_ros/create_timer_ros.h>
   #include <tf2_ros/message_filter.h>
   #include <tf2_ros/transform_listener.h>
   #ifdef TF2_CPP_HEADERS
     #include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
   #else
     #include <tf2_geometry_msgs/tf2_geometry_msgs.h>
   #endif

> Secondly, the persistent data. There need to be persistent instances of ``tf2_ros::Buffer``, ``tf2_ros::TransformListener`` and ``tf2_ros::MessageFilter``.

Reviewer: ```suggestion
Second, there needs to be persistent instances of ``tf2_ros::Buffer``, ``tf2_ros::TransformListener`` and ``tf2_ros::MessageFilter``.
```

Author: I changed it.

.. code-block:: C++

   std::string target_frame_;
   std::shared_ptr<tf2_ros::Buffer> tf2_buffer_;
   std::shared_ptr<tf2_ros::TransformListener> tf2_listener_;
   message_filters::Subscriber<geometry_msgs::msg::PointStamped> point_sub_;
   std::shared_ptr<tf2_ros::MessageFilter<geometry_msgs::msg::PointStamped>> tf2_filter_;

> Thirdly, the constructor. When starting up the ROS 2 ``message_filters::Subscriber`` must be initialized with the topic. And the ``tf2_ros::MessageFilter`` must be initialized with that ``Subscriber`` object.

Reviewer: ```suggestion
Third, the ROS 2 ``message_filters::Subscriber`` must be initialized with the topic.
And the ``tf2_ros::MessageFilter`` must be initialized with that ``Subscriber`` object.
```

Author: I corrected it, and split this paragraph into two.

> The other arguments of note in the ``MessageFilter`` constructor are the ``target_frame`` and the callback function. The target frame is the frame into which it will make sure ``canTransform`` will succeed. And the callback function is the function that will be called when the data is ready.

Reviewer: ```suggestion
The other arguments of note in the ``MessageFilter`` constructor are the ``target_frame`` and the callback function.
The target frame is the frame into which it will make sure ``canTransform`` will succeed.
And the callback function is the function that will be called when the data is ready.
```

Author: I split this paragraph into 3 paragraphs.

     this->get_parameter("target_frame", target_frame_);

     typedef std::chrono::duration<int> seconds_type;
     seconds_type buffer_timeout(1);

     tf2_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
     // Create the timer interface before call to waitForTransform,
     // to avoid a tf2_ros::CreateTimerInterfaceException exception
     auto timer_interface = std::make_shared<tf2_ros::CreateTimerROS>(
       this->get_node_base_interface(),
       this->get_node_timers_interface());
     tf2_buffer_->setCreateTimerInterface(timer_interface);
     tf2_listener_ =
       std::make_shared<tf2_ros::TransformListener>(*tf2_buffer_);

     point_sub_.subscribe(this, "/turtle3/turtle_point_stamped");
     tf2_filter_ = std::make_shared<tf2_ros::MessageFilter<geometry_msgs::msg::PointStamped>>(
       point_sub_, *tf2_buffer_, target_frame_, 100, this->get_node_logging_interface(),
       this->get_node_clock_interface(), buffer_timeout);
     // Register a callback with tf2_ros::MessageFilter to be called when transforms are available
     tf2_filter_->registerCallback(&PoseDrawer::msgCallback, this);
   }

> And lastly, the callback method. Once the data is ready, just call ``tf2_buffer_->transform`` and print outputs to the console.

Reviewer: ```suggestion
And last, the callback method will call ``tf2_buffer_->transform`` when the data is ready and print output to the console.
```

Author: I changed these as your suggestion.

.. code-block:: C++

   private:
   void msgCallback(const geometry_msgs::msg::PointStamped::SharedPtr point_ptr)
   {
     geometry_msgs::msg::PointStamped point_out;
     try {
       tf2_buffer_->transform(*point_ptr, point_out, target_frame_);
       RCLCPP_INFO(
         this->get_logger(), "Point of turtle3 in frame of turtle1: x:%f y:%f z:%f\n",
         point_out.point.x,
         point_out.point.y,
         point_out.point.z);
     } catch (tf2::TransformException & ex) {
       RCLCPP_WARN(
         // Print exception which was caught
         this->get_logger(), "Failure %s\n", ex.what());
     }
   }

2.2 Build the package
~~~~~~~~~~~~~~~~~~~~~

Before building the package ``learning_tf2_cpp``, please add two another dependencies in the ``package.xml`` file of this package:

.. code-block:: xml

   <depend>message_filters</depend>
   <depend>tf2_geometry_msgs</depend>

And in the ``CMakeLists.txt`` file, add two lines below the existing dependencies:

.. code-block:: console

   find_package(message_filters REQUIRED)
   find_package(tf2_geometry_msgs REQUIRED)

> Most import, add these lines below the dependencies:

Reviewer: ```suggestion
Most importantly, add these lines below the dependencies:
```

Author: I corrected this syntax error.

   ament_target_dependencies(
     turtle_tf2_message_filter
     geometry_msgs
     message_filters
     rclcpp
     tf2
     tf2_geometry_msgs
     tf2_ros
   )

Finally, add the ``install(TARGETS…)`` section (below other existing nodes) so ``ros2 run`` can find your executable:

.. code-block:: console

   install(TARGETS
     turtle_tf2_message_filter
     DESTINATION lib/${PROJECT_NAME})

Now open a new terminal, navigate to the root of your workspace, and rebuild the package with command:

.. tabs::

  .. group-tab:: Linux

    .. code-block:: console

       colcon build --packages-select learning_tf2_cpp

  .. group-tab:: macOS

    .. code-block:: console

       colcon build --packages-select learning_tf2_cpp

  .. group-tab:: Windows

    .. code-block:: console

       colcon build --merge-install --packages-select learning_tf2_cpp

Running and results
-------------------

First we need to run several nodes (including the broadcaster node of PointStamped messages) by launching the launch file ``turtle_tf2_sensor_message.launch.py``:

.. code-block:: console

   ros2 launch learning_tf2_py turtle_tf2_sensor_message.launch.py

> This will bring up the ``turtlesim`` window with two turtles, where ``turtle3`` is moving along a circle, while ``turtle1`` isn't moving at first. But you can run the ``turtle_teleop_key`` node in another terminal to drive ``turtle1`` to move:

Reviewer: ```suggestion
This will bring up the ``turtlesim`` window with two turtles, where ``turtle3`` is moving along a circle, while ``turtle1`` isn't moving at first.
But you can run the ``turtle_teleop_key`` node in another terminal to drive ``turtle1`` to move:
```

Author: I split it into 2 paragraphs.

.. code-block:: console

   ros2 run turtlesim turtle_teleop_key

.. image:: images/turtlesim_messagefilter.png

Now if you echo the topic ``turtle3/turtle_point_stamped``:

.. code-block:: console

   ros2 topic echo /turtle3/turtle_point_stamped

> Then there will have outputs like this:

Reviewer: ```suggestion
Then there will be output like this:
```

Author: I corrected this error.

.. code-block:: console

   header:
     stamp:
       sec: 1629877510
       nanosec: 902607040
     frame_id: world
   point:
     x: 4.989276885986328
     y: 3.073937177658081
     z: 0.0
   ---
   header:
     stamp:
       sec: 1629877510
       nanosec: 918389395
     frame_id: world
   point:
     x: 4.987966060638428
     y: 3.089883327484131
     z: 0.0
   ---
   header:
     stamp:
       sec: 1629877510
       nanosec: 934186680
     frame_id: world
   point:
     x: 4.986400127410889
     y: 3.105806589126587
     z: 0.0
   ---

When the demo is running, open another terminal and run the message filter/listener node:

.. code-block:: console

   ros2 run learning_tf2_cpp turtle_tf2_message_filter

> If it's running right you should see streaming data like this:

Reviewer: ```suggestion
If it's running correctly you should see streaming data like this:
```

Author: I changed that.

.. code-block:: console

   [INFO] [1630016162.006173900] [turtle_tf2_pose_drawer]: Point of turtle3 in frame of turtle1: x:-6.493231 y:-2.961614 z:0.000000

   [INFO] [1630016162.006291983] [turtle_tf2_pose_drawer]: Point of turtle3 in frame of turtle1: x:-6.472169 y:-3.004742 z:0.000000

   [INFO] [1630016162.006326234] [turtle_tf2_pose_drawer]: Point of turtle3 in frame of turtle1: x:-6.479420 y:-2.990479 z:0.000000

   [INFO] [1630016162.006355644] [turtle_tf2_pose_drawer]: Point of turtle3 in frame of turtle1: x:-6.486441 y:-2.976102 z:0.000000

Summary
-------

> In this tutorial you learned how to use sensor data/messages in tf2. Specifically speaking, you learned how to publish ``PointStamped`` messages on a topic, and how to listen to the topic and transform the frame of ``PointStamped`` messages with ``tf2_ros::MessageFilter``.

Reviewer: ```suggestion
In this tutorial you learned how to use sensor data/messages in tf2.
Specifically speaking, you learned how to publish ``PointStamped`` messages on a topic, and how to listen to the topic and transform the frame of ``PointStamped`` messages with ``tf2_ros::MessageFilter``.
```

Author: I split it into 2 paragraphs.

               executable='turtle_tf2_broadcaster',
               name='broadcaster2',
               parameters=[
                   {'turtlename': 'turtle3'}
               ]
           ),
           Node(
               package='turtle_tf2_py',
               executable='turtle_tf2_message_broadcaster',
               name='message_broadcaster',
           ),
       ])

1.3 Add an entry point and build the pakcage
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Don't forget to add the executable in the ``setup.py`` file of the package:

.. code-block:: python

   'console_scripts': [
       ...
       'turtle_tf2_message_broadcaster = learning_tf2_py.turtle_tf2_message_broadcaster:main',
   ],

And then we can build the package:

.. tabs::

  .. group-tab:: Linux

    .. code-block:: console

       colcon build --packages-select learning_tf2_py

  .. group-tab:: macOS

    .. code-block:: console

       colcon build --packages-select learning_tf2_py

  .. group-tab:: Windows

    .. code-block:: console

       colcon build --merge-install --packages-select learning_tf2_py

2 Writing the message filter/listener node
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Now, to get the streaming ``PointStamped`` data of ``turtle3`` in the frame of ``turtle1`` reliably, we will create the source file of message filter/listener node at first.

Reviewer: ```suggestion
Now, to get the streaming ``PointStamped`` data of ``turtle3`` in the frame of ``turtle1`` reliably, we will create the source file of the message filter/listener node.
```

Author: Ok, I changed that.

.. code-block:: C++

   private:
   void msgCallback(const geometry_msgs::msg::PointStamped::SharedPtr point_ptr)
   {
     geometry_msgs::msg::PointStamped point_out;
     try {
       tf2_buffer_->transform(*point_ptr, point_out, target_frame_);
       RCLCPP_INFO(
         this->get_logger(), "Point of turtle3 in frame of turtle1: x:%f y:%f z:%f\n",
         point_out.point.x,
         point_out.point.y,
         point_out.point.z);
     } catch (tf2::TransformException & ex) {
       RCLCPP_WARN(
         // Print exception which was caught
         this->get_logger(), "Failure %s\n", ex.what());
     }
   }

2.2 Build the package
~~~~~~~~~~~~~~~~~~~~~

Before building the package ``learning_tf2_cpp``, please add two another dependencies in the ``package.xml`` file of this package:

.. code-block:: xml

   <depend>message_filters</depend>
   <depend>tf2_geometry_msgs</depend>

And in the ``CMakeLists.txt`` file, add two lines below the existing dependencies:

.. code-block:: console

   find_package(message_filters REQUIRED)
   find_package(tf2_geometry_msgs REQUIRED)

> Most importly, add these lines below the dependencies:

Reviewer: ```suggestion
Most importantly, add these lines below the dependencies:
```

Author: Yes, another error, thank you very much, I corrected it just now.

.. _UsingStampedDatatypesWithTf2RosMessageFilter:

Using stamped datatypes with tf2_ros::MessageFilter
===================================================

> **Goal:** Learn how to use tf2_ros::MessageFilter to process stamped datatypes.

Reviewer 2: ```suggestion
**Goal:** Learn how to use ``tf2_ros::MessageFilter`` to process stamped datatypes.
```

Author: I did it just now.

> **Tutorial level:** Intermediate

Reviewer 2: You can mark this tutorial as Advanced I think

Author: Awesome, I think it's Advanced, so I changed that.

Author: And I move the image to the images sub-folder as you did, and changed the image path accordingly in the rst file.

**Time:** 10 minutes

.. contents:: Contents
   :depth: 3
   :local:

Background
----------

This tutorial explains how to use sensor data with tf2. Some real-world examples of sensor data are:

    * cameras, both mono and stereo

    * laser scans

Suppose that a new turtle named ``turtle3`` is created and it doesn't have good odometry, but there is an overhead camera tracking its position and publishing it as a ``PointStamped`` message in relation to the ``world`` frame.

``turtle1`` wants to know where ``turtle3`` is compared to itself.

> To do this ``turtle1`` must listen to the topic where ``turtle3``'s pose is being published, wait until transforms into the desired frame are ready, and then do it's operations.

Reviewer 2: ```suggestion
To do this ``turtle1`` must listen to the topic where ``turtle3``'s pose is being published, wait until transforms into the desired frame are ready, and then do its operations.
```

Author: Yes, you're right, it's a syntax error, I changed it.

> To make this easier the ``tf2_ros::MessageFilter`` is very useful. The ``tf2_ros::MessageFilter`` will take a subscription to any ROS 2 message with a header and cache it until it is possible to transform it into the target frame.

Reviewer 2: Do you think we can paraphrase that first sentence? I think it would be best to clarify what is it making easier and how

Author: Yes, I think the "this" in the "To make this easier" sentence refers to the sentence "To do this "turtle1" must listen to ...", so I merged these two paragraphs into one, and I think that can clarify what is it making easier and how.

Setting the example
-------------------

1 Write the broadcaster node of PointStamped messages
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> For this tutorial we will set up a demo application which has a node (in Python) to broadcast the ``PointStamped`` messages of ``turtle3``.

Reviewer 2: I think adding that we are publishing a ``PointStamped`` **position** messages of ``turtle3`` will make it clear

Author: I changed that.

  .. group-tab:: Windows

    In a Windows command line prompt:

    .. code-block:: console

       curl -sk https://raw.githubusercontent.com/ros/geometry_tutorials/ros2/turtle_tf2_py/turtle_tf2_py/turtle_tf2_message_broadcaster.py -o turtle_tf2_message_broadcaster.py

    Or in powershell:

    .. code-block:: console

       curl https://raw.githubusercontent.com/ros/geometry_tutorials/ros2/turtle_tf2_py/turtle_tf2_py/turtle_tf2_message_broadcaster.py -o turtle_tf2_message_broadcaster.py

Open the file using your preferred text editor.

.. code-block:: python

   # Copyright 2021 Open Source Robotics Foundation, Inc.
   #
   # Licensed under the Apache License, Version 2.0 (the "License");
   # you may not use this file except in compliance with the License.
   # You may obtain a copy of the License at
   #
   #     http://www.apache.org/licenses/LICENSE-2.0
   #
   # Unless required by applicable law or agreed to in writing, software
   # distributed under the License is distributed on an "AS IS" BASIS,
   # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   # See the License for the specific language governing permissions and

>    # limitations under the License.

Reviewer 2: Usually we omit copyright notices in code snippets

Author: Ok, I deleted all compyright notices in the code snippets.

           pass

       rclpy.shutdown()

1.1 Examine the code
~~~~~~~~~~~~~~~~~~~~

Now let's take a look at the code.
Firstly, in the ``on_timer`` callback function, we spawn the ``turtle3`` by asynchronously calling the ``Spawn`` service of ``turtlesim``, and initialize its position at (4, 2, 0), when the turtle spawning service is ready.

.. code-block:: python

   # Initialize request with turtle name and coordinates
   # Note that x, y and theta are defined as floats in turtlesim/srv/Spawn
   request = Spawn.Request()
   request.name = 'turtle3'
   request.x = float(4)
   request.y = float(2)
   request.theta = float(0)
   Call request
   self.result = self.spawner.call_async(request)

Afterward, the node publishes the topic ``turtle3/cmd_vel``, topic ``turtle3/turtle_point_stamped``, and subscribes to topic ``turtle3/pose`` and runs callback function ``handle_turtle_pose`` on every incoming message.

.. code-block:: python

   self.vel_pub = self.create_publisher(Twist, '/turtle3/cmd_vel', 10)
   self.sub = self.create_subscription(Pose, '/turtle3/pose', self.handle_turtle_pose, 10)
   self.pub = self.create_publisher(PointStamped, '/turtle3/turtle_point_stamped', 10)

> Finally, in the callback function ``handle_turtle_pose``, we initialize the ``Twist`` messages of ``turtle3`` and publish them, which will make the ``turtle3`` moving along a circle, then we fill up the ``PointStamped`` messages of ``turtle3`` with incoming ``Pose`` messages of itself and publish it.

Reviewer 2: ```suggestion
Finally, in the callback function ``handle_turtle_pose``, we initialize the ``Twist`` messages of ``turtle3`` and publish them, which will make the ``turtle3`` move along a circle. Then we fill up the ``PointStamped`` messages of ``turtle3`` with incoming ``Pose`` messages and publish them.
```

Author: You're very careful to help me find these errors. I changed them, thank you.

.. code-block:: python

   vel_msg = Twist()
   vel_msg.linear.x = 1.0
   vel_msg.angular.z = 1.0
   self.vel_pub.publish(vel_msg)

   ps = PointStamped()
   ps.header.stamp = self.get_clock().now().to_msg()
   ps.header.frame_id = 'world'
   ps.point.x = msg.x
   ps.point.y = msg.y
   ps.point.z = 0.0
   self.pub.publish(ps)

1.2 Write the launch file
~~~~~~~~~~~~~~~~~~~~~~~~~

In order to run this demo, we need to create a launch file ``turtle_tf2_sensor_message.launch.py`` in the ``launch`` subdirectory of package ``learning_tf2_py``:

.. code-block:: python

   # Copyright 2021 Open Source Robotics Foundation, Inc.
   #
   # Licensed under the Apache License, Version 2.0 (the "License");
   # you may not use this file except in compliance with the License.
   # You may obtain a copy of the License at
   #
   #     http://www.apache.org/licenses/LICENSE-2.0
   #
   # Unless required by applicable law or agreed to in writing, software
   # distributed under the License is distributed on an "AS IS" BASIS,
   # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   # See the License for the specific language governing permissions and

>    # limitations under the License.

Reviewer 2: same for this copyright

   from launch import LaunchDescription
   from launch.actions import DeclareLaunchArgument
   from launch_ros.actions import Node

   def generate_launch_description():
       return LaunchDescription([
           DeclareLaunchArgument(
               'target_frame', default_value='turtle1',
               description='Target frame name.'
           ),
           Node(
               package='turtlesim',
               executable='turtlesim_node',
               name='sim',
               output='screen'
           ),
           Node(
               package='turtle_tf2_py',
               executable='turtle_tf2_broadcaster',
               name='broadcaster1',
               parameters=[
                   {'turtlename': 'turtle1'}
               ]
           ),
           Node(
               package='turtle_tf2_py',
               executable='turtle_tf2_broadcaster',
               name='broadcaster2',
               parameters=[
                   {'turtlename': 'turtle3'}
               ]
           ),
           Node(
               package='turtle_tf2_py',

>                executable='turtle_tf2_message_broadcaster',

Reviewer 2: Is this correct that the executable name here  and on line 289 are different? First is `turtle_tf2_message_broadcaster` and the second is `turtle_tf2_msg_broadcaster`. I think you need to update one in the line 289

Author: You're right, I modified "turtle_tf2_msg_broadercaster" in line 289 to "turtle_tf2_message_broadcaster".

               name='message_broadcaster',
           ),
       ])

1.3 Add an entry point and build the pakcage
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Don't forget to add the executable in the ``setup.py`` file of the package:

.. code-block:: python

   'console_scripts': [

>        'turtle_tf2_msg_broadcaster = learning_tf2_py.turtle_tf2_message_broadcaster:main',

Reviewer 2: ```suggestion
       ...
       'turtle_tf2_msg_broadcaster = learning_tf2_py.turtle_tf2_message_broadcaster:main',
```
I think it would be better to add three dots here to indicate that we are adding that node and not removing previous ones

Author: Yes, you're right. I added three dots here.

     }

   private:
     void msgCallback(const geometry_msgs::msg::PointStamped::SharedPtr point_ptr)
     {
       geometry_msgs::msg::PointStamped point_out;
       try {
         tf2_buffer_->transform(*point_ptr, point_out, target_frame_);
         RCLCPP_INFO(
           this->get_logger(), "Point of turtle3 in frame of turtle1: x:%f y:%f z:%f\n",
           point_out.point.x,
           point_out.point.y,
           point_out.point.z);
       } catch (tf2::TransformException & ex) {
         RCLCPP_WARN(
           // Print exception which was caught
           this->get_logger(), "Failure %s\n", ex.what());
       }
     }
     std::string target_frame_;
     std::shared_ptr<tf2_ros::Buffer> tf2_buffer_;
     std::shared_ptr<tf2_ros::TransformListener> tf2_listener_;
     message_filters::Subscriber<geometry_msgs::msg::PointStamped> point_sub_;
     std::shared_ptr<tf2_ros::MessageFilter<geometry_msgs::msg::PointStamped>> tf2_filter_;
   };

   int main(int argc, char * argv[])
   {
     rclcpp::init(argc, argv);
     rclcpp::spin(std::make_shared<PoseDrawer>());
     rclcpp::shutdown();
     return 0;
   }

> You can download this code directly from file ``turtle_tf2_message_filter.cpp`` in the ``src`` subdirectory of the ``learning_tf2_cpp`` :ref:`package <WritingATf2StaticBroadcasterCpp>`.

Reviewer 2: I think it would be better to add wget links in a similar manner before the code itself and remove that part

Author: Ok, I added wget links before the code itself.

.. code-block:: C++

   #include <geometry_msgs/msg/point_stamped.hpp>
   #include <message_filters/subscriber.h>

   #include <rclcpp/rclcpp.hpp>
   #include <tf2_ros/buffer.h>
   #include <tf2_ros/create_timer_ros.h>
   #include <tf2_ros/message_filter.h>
   #include <tf2_ros/transform_listener.h>
   #ifdef TF2_CPP_HEADERS
     #include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
   #else
     #include <tf2_geometry_msgs/tf2_geometry_msgs.h>
   #endif

Secondly, the persistent data. There need to be persistent instances of ``tf2_ros::Buffer``, ``tf2_ros::TransformListener`` and ``tf2_ros::MessageFilter``.

.. code-block:: C++

   std::string target_frame_;
   std::shared_ptr<tf2_ros::Buffer> tf2_buffer_;
   std::shared_ptr<tf2_ros::TransformListener> tf2_listener_;
   message_filters::Subscriber<geometry_msgs::msg::PointStamped> point_sub_;
   std::shared_ptr<tf2_ros::MessageFilter<geometry_msgs::msg::PointStamped>> tf2_filter_;

Thirdly, the constructor. When starting up the ROS 2 ``message_filters::Subscriber`` must be initialized with the topic. And the ``tf2_ros::MessageFilter`` must be initialized with that ``Subscriber`` object.

> The other arguments of note in the ``MessageFilter`` constructor are the ``target_frame`` and callback function. The target frame is the frame into which it will make sure ``canTransform`` will succeed. And the callback function is the function which will be called when the data is ready.

Reviewer 2: ```suggestion
The other arguments of note in the ``MessageFilter`` constructor are the ``target_frame`` and the callback function. The target frame is the frame into which it will make sure ``canTransform`` will succeed. And the callback function is the function that will be called when the data is ready.
```

Author: Yes, my english is not good, so I actually don't know what differences between "which" and "that" when they are used in clause. Anyway, I changed it as your suggestion.

Reviewer 2: Mine is neither, don't worry 😂
We are all learning in the process

     this->get_parameter("target_frame", target_frame_);

     typedef std::chrono::duration<int> seconds_type;
     seconds_type buffer_timeout(1);

     tf2_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
     // Create the timer interface before call to waitForTransform,
     // to avoid a tf2_ros::CreateTimerInterfaceException exception
     auto timer_interface = std::make_shared<tf2_ros::CreateTimerROS>(
       this->get_node_base_interface(),
       this->get_node_timers_interface());
     tf2_buffer_->setCreateTimerInterface(timer_interface);
     tf2_listener_ =
       std::make_shared<tf2_ros::TransformListener>(*tf2_buffer_);

     point_sub_.subscribe(this, "/turtle3/turtle_point_stamped");
     tf2_filter_ = std::make_shared<tf2_ros::MessageFilter<geometry_msgs::msg::PointStamped>>(
       point_sub_, *tf2_buffer_, target_frame_, 100, this->get_node_logging_interface(),
       this->get_node_clock_interface(), buffer_timeout);
     // Register a callback with tf2_ros::MessageFilter to be called when transforms are available
     tf2_filter_->registerCallback(&PoseDrawer::msgCallback, this);
   }

> And lastly, the callback method. Once the data is ready, just call ``tf2_buffer_->transform`` and print to screen for the tutorial.

Reviewer 2: We can rephrase that as "print output to the console"

Author: I changed that, too.

.. code-block:: console

   install(TARGETS
     turtle_tf2_message_filter
     DESTINATION lib/${PROJECT_NAME})

Now open a new terminal, navigate to the root of your workspace, and rebuild the package with command:

.. tabs::

  .. group-tab:: Linux

    .. code-block:: console

       colcon build --packages-select learning_tf2_cpp

  .. group-tab:: macOS

    .. code-block:: console

       colcon build --packages-select learning_tf2_cpp

  .. group-tab:: Windows

    .. code-block:: console

       colcon build --merge-install --packages-select learning_tf2_cpp

Running and results
-------------------

First we need to run several nodes (including the broadcaster node of PointStamped messages) by launching the launch file ``turtle_tf2_sensor_message.launch.py``:

.. code-block:: console

   ros2 launch learning_tf2_py turtle_tf2_sensor_message.launch.py

This will bring up the ``turtlesim`` window with two turtles, where ``turtle3`` is moving along a circle, while ``turtle1`` isn't moving at first. But you can run the ``turtle_teleop_key`` node in another terminal to drive ``turtle1`` to move:

.. code-block:: console

   ros2 run turtlesim turtle_teleop_key

> .. image:: turtlesim_messagefilter.png

Reviewer 2: Can you crop the image to remove that window header? So that it matches all other pictures in tf2 tutorials

Author: Ok, I cropped the image just now.

     z: 0.0
   ---
   header:
     stamp:
       sec: 1629877510
       nanosec: 918389395
     frame_id: world
   point:
     x: 4.987966060638428
     y: 3.089883327484131
     z: 0.0
   ---
   header:
     stamp:
       sec: 1629877510
       nanosec: 934186680
     frame_id: world
   point:
     x: 4.986400127410889
     y: 3.105806589126587
     z: 0.0
   ---

When the demo is running, open another terminal and run the message filter/listener node:

.. code-block:: console

   ros2 run learning_tf2_cpp turtle_tf2_message_filter

If it's running right you should see streaming data like this:

.. code-block:: console

   [INFO] [1630016162.006173900] [turtle_tf2_pose_drawer]: Point of turtle3 in frame of turtle1: x:-6.493231 y:-2.961614 z:0.000000

   [INFO] [1630016162.006291983] [turtle_tf2_pose_drawer]: Point of turtle3 in frame of turtle1: x:-6.472169 y:-3.004742 z:0.000000

   [INFO] [1630016162.006326234] [turtle_tf2_pose_drawer]: Point of turtle3 in frame of turtle1: x:-6.479420 y:-2.990479 z:0.000000

   [INFO] [1630016162.006355644] [turtle_tf2_pose_drawer]: Point of turtle3 in frame of turtle1: x:-6.486441 y:-2.976102 z:0.000000

Summary
-------

> In this tutorial you learned how to use sensor data/messages in tf2. Specifically speaking, you learned how to publish PointStamped messages on a topic, and how to listen to the topic and transform the frame of PointStamped messages with tf2_ros::MessageFilter.

Reviewer 2: ```suggestion
In this tutorial you learned how to use sensor data/messages in tf2. Specifically speaking, you learned how to publish ``PointStamped`` messages on a topic, and how to listen to the topic and transform the frame of ``PointStamped`` messages with ``tf2_ros::MessageFilter``.
```

Author: I did it.
